File: LeetCode/C++/LeetCode0200_NumberOfIslands.cpp
// LeetCode0200_NumberOfIslands.cpp
#define VISIT 1
#define UNVISIT 0

typedef struct _Position {
	int row;
	int col;
} Position;

class Solution {
public:
	int rowSize, colSize;
	vector<vector<int>> visited;

	void BFS(vector<vector<char>>& grid, Position target) {
		queue<Position> q;

		q.push(target);
		while (! q.empty()) {
			Position curr = q.front(); q.pop();

			Position surr[4] = {
				{ curr.row - 1, curr.col }, // Up
				{ curr.row, curr.col + 1 }, // Right
				{ curr.row + 1, curr.col }, // Down
				{ curr.row, curr.col - 1 }  // Left
			};

			for (int i = 0; i < 4; i++) {
				int currRow = surr[i].row, currCol = surr[i].col;
				if ((currRow >= 0 && currCol >= 0) && (currRow < rowSize && currCol < colSize))
					if ( visited[currRow][currCol] == UNVISIT ) {
						visited[currRow][currCol] = VISIT;
						if ( grid[currRow][currCol] == '1' )
							q.push( { currRow, currCol } );
					}
			}			
		}
	}

	int numIslands(vector<vector<char>>& grid) {
		if (grid.size() == 0) return 0;
		rowSize = grid.size();
		colSize = grid[0].size();

		visited = vector<vector<int>> (rowSize, vector<int> (colSize));

		int count = 0;

		for (int i = 0; i < rowSize; i++)
			for (int j = 0; j < colSize; j++)
				if (visited[i][j] == UNVISIT) {
					visited[i][j] = VISIT;
					if (grid[i][j] == '1') {
						BFS(grid, { i, j });
						count++;
					}
				}

		return count;
	}
};

File: LeetCode/Python3/LeetCode0200_NumberOfIslands.py
class Solution:
	def numIslands( self, grid: List[List[str]] ) -> int:
		if not grid:
			return 0

		maxRow, maxCol = len( grid ), len( grid[0] )
		visited = [ [ False ] * maxCol for _ in range( maxRow ) ]
		
		count = 0
		for i in range( maxRow ):
			for j in range( maxCol ):
				if not visited[i][j]:
					visited[i][j] = True

					if grid[i][j] == '1':
						q = [ ( i, j ) ]
						while q:
							currRow, currCol = q.pop( 0 )

							nextPos = [
								( currRow - 1, currCol ), # Up
								( currRow, currCol - 1 ), # Left
								( currRow, currCol + 1 ), # Right
								( currRow + 1 , currCol ), # Bottom
							]

							for nextRow, nextCol in nextPos:
								if ( ( -1 < nextRow < maxRow ) and ( -1 < nextCol < maxCol ) ):
									if not visited[nextRow][nextCol]:
										visited[nextRow][nextCol] = True
										
										if grid[nextRow][nextCol] == '1':
											q.append( ( nextRow, nextCol ) )

						count += 1

		return count
